### ISSUE TITLE
Send custom messages from JS to Python

### ISSUE THREAD

**Issue by domoritz** [manzt/anywidget#54]
https://anywidget.dev/en/jupyter-widgets-the-good-parts/#2-custom-messages explains how to send custom messages to the frontend but is there an equivalent to send data to the backend? I know traitlets that are synchronized can be used for this but I wonder whether there is a way to send messages that I don't need to synchronize back to the frontend.

**Comment by manzt** [comment]
Yes! You can use the `DOMWidgetModel.send` method in the frontend to send messages from JS to Python that don't need synchronization. This works for any Jupyter Widget, but you can try it out easily with anywidget:

```python
import anywidget

class CustomMessageWidget(anywidget.AnyWidget):
    _esm = """
    export function render(view) {
        setInterval(() => {
            view.model.send({ foo: "bar" });
        }, 2000);
    }
    """
        
widget = CustomMessageWidget()

@widget.on_msg
def do_something(instance: CustomMessageWidget, data: dict, buffers: list):
    print(f"{instance=}, {data=}, {buffers=}")
    
widget
```

https://user-images.githubusercontent.com/24403730/216881461-9e16d06a-5c6b-46f4-9b17-95dfbda6f1a9.mov

It's worth noting that you _should_ be able to send not only JSON-serializable data, but also binary data from the client to Python (that's what the `buffers` should be populated with in a custom message hanlder), but there is currently a [bug preventing this](https://github.com/jupyter-widgets/ipywidgets/pull/3689). I will implement support in **anywidget** in the meantime.

**Comment by manzt** [comment]
It's worth noting that I haven't found a nice pattern in this framework to block/wait for a response from the front end for different messages. I.e.,

```python

class MyWidget(anywidget.AnyWidget):
    # ...
    def save_png(self):
        self.send({ "export-png" })
        # somehow wait for response ...
```

**Comment by manzt** [comment]
> However, I want to go the other way around and request something from the Python side. I came up with a hack where messages get an ID and the frontend would keep a map of promises that it resolves when the response with a particular id arrives.

Ha, yes I've come up with something very similar for this!

> One thing I am not getting yet is how to listen to messages in the class rather than a specific instance (@widget.on_msg only works when you already have an instance of the widget).

You can add an `on_msg` handler in the constructor of your widget:

```python
class MyWidget(anywidget.AnyWidget):
    _esm = " .... "

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.on_msg(_handle_msg)

    def _handle_msg(self, instance, data, buffers):
        ...

```

**Comment by domoritz** [comment]
I think it's

```python
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.on_msg(self._handle_msg)

    def _handle_msg(self, message):
        print(f"{message=}")
```

no?

**Comment by manzt** [comment]
Ah! Actually, `_handle_msg` is implemented on the base class, so this example breaks all the other functionality implemented by the widget. Pick a different name:

```python
import anywidget

class MyWidget(anywidget.AnyWidget):
    _esm = """
    export function render(view) {
      let btn = Object.assign(document.createElement("button"), { innerText: "Click me" });
      btn.addEventListener("click", () => { view.model.send("ping") });
      view.el.appendChild(btn);
    }
    """

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.on_msg(self._handle_custom_msg)

    def _handle_custom_msg(self, data, buffers):
        print(f"{data=}, {buffers=}")
```

**Comment by domoritz** [comment]
Ahh, good catch. Thanks!
